plugin/evm/statesync/queue.go

```diff
@@ -0,0 +1,92 @@
+// (c) 2019-2020, Ava Labs, Inc.
+//
+// This file is although named queue.go, is not the same as the eth/downloader/queue.go file.
+// Instead, this is simply the protected array for downloader (yeah not even a queue).
+// It might be helpful to keep this in case we move computation here
+//
+// It is distributed under a license compatible with the licensing terms of the
+// original code from which it is derived.
+
+package statesync
+
+import (
+	"fmt"
+	"sync"
+)
+
+type Executable interface {
+	ExitQueue() error
+}
+
+type Queue[K Executable] struct {
+	buffer  []*K
+	l       sync.RWMutex
+	nextPos int
+	compare func(*K, *K) int
+	closed  bool
+}
+
+func NewQueue[K Executable](size int, compare func(*K, *K) int) *Queue[K] {
+	return &Queue[K]{
+		buffer:  make([]*K, size),
+		compare: compare,
+	}
+}
+
+func (q *Queue[K]) Insert(h *K) error {
+	q.l.Lock()
+	defer q.l.Unlock()
+
+	if q.nextPos >= len(q.buffer) {
+		return fmt.Errorf("queue is full, cannot insert")
+	}
+
+	q.buffer[q.nextPos] = h
+	q.nextPos++
+
+	return nil
+}
+
+func (q *Queue[K]) Flush(max *K, close bool) error {
+	q.l.Lock()
+	defer q.l.Unlock()
+
+	newBuffer := make([]*K, len(q.buffer))
```

Comment:
seems this can be done without allocating a new queue.
1. find the position to evict,
2. evict all stuff after that,
3. set nextPos to the found position